applications/OptimizationApplication/custom_utilities/filtering/explicit_vertex_morphing_filter.cpp

```diff
@@ -0,0 +1,342 @@
+        << "Provided contaienr expression = " << rContainerExpression << ".\n";
+
+    KRATOS_ERROR_IF_NOT(&rContainerExpression.GetModelPart() == &mrModelPart)
+        << "Filter radius container expression model part and filter model part mismatch."
+        << "\n\tFilter = " << *this
+        << "\n\tContainerExpression = " << rContainerExpression;
+
+    mpFilterRadiusContainer = rContainerExpression.Clone();
+}
+
+template<class TContainerType>
+void ExplicitVertexMorphingFilter<TContainerType>::ExplicitVertexMorphingFilter::Update()
+{
+    KRATOS_TRY
+
+    BuiltinTimer timer;
+
+    KRATOS_INFO("ExplicitVertexMorphingFilter") << "Creating search tree to perform mapping..." << std::endl;
+
+    const auto& r_container = ExplicitVertexMorphingFilterHelperUtilities::GetContainer<TContainerType>(mrModelPart);
+
+    if (mEntityPointVector.size() != r_container.size()) {
+        mEntityPointVector.resize(r_container.size());
+    }
+
+    // now fill the points vector
+    IndexPartition<IndexType>(r_container.size()).for_each([&](const IndexType Index) {
+        mEntityPointVector[Index] = Kratos::make_shared<EntityPoint<EntityType>>(*(r_container.begin() + Index), Index);
+    });
+
+    mpSearchTree =  Kratos::make_shared<ExplicitVertexMorphingFilter::KDTree>(mEntityPointVector.begin(), mEntityPointVector.end(), mBucketSize);
+
+    if constexpr(std::is_same_v<TContainerType, ModelPart::NodesContainerType>) {
+        const auto& r_nodes = mrModelPart.Nodes();
+        const auto& r_conditions = mrModelPart.Conditions();
+        const IndexType number_of_conditions = r_conditions.size();
+
+        const auto& r_elements = mrModelPart.Elements();
+        const IndexType number_of_elements = r_elements.size();
+
+        KRATOS_ERROR_IF(number_of_conditions > 0 && number_of_elements > 0) << mrModelPart.FullName()
+            << " has both elements and conditions. Nodal mapping model parts should only have either conditions or elements.\n";
+
+        if (number_of_conditions > 0) {
+            mpNodalDomainSizeExpression = ExplicitVertexMorphingFilterHelperUtilities::GetNodalDomainSizeExpression(r_conditions, r_nodes);
+        } else if (number_of_elements > 0) {
+            mpNodalDomainSizeExpression = ExplicitVertexMorphingFilterHelperUtilities::GetNodalDomainSizeExpression(r_elements, r_nodes);
+        } else {
+            KRATOS_ERROR << "Nodal mapping requires atleast either conditions or elements to be present in "
+                         << mrModelPart.FullName() << ".\n";
+        }
+    }
+
+    KRATOS_INFO("ExplicitVertexMorphingFilter") << "Search tree created in: " << timer.ElapsedSeconds() << " s" << std::endl;
+
+    KRATOS_CATCH("");
+}
+
+template<class TContainerType>
+template<class TWeightIntegrationType>
+ContainerExpression<TContainerType> ExplicitVertexMorphingFilter<TContainerType>::GenericFilterField(const ContainerExpression<TContainerType>& rContainerExpression) const
```

Comment:
If we have elementalexpression do you do kdtree search on nodes or element centers ?

Reply:
In this case, Kd tree search is done based on the element centeres (ConitionsExpressions -> Condition centeres)